pyteomics/proforma.py

```diff
@@ -0,0 +1,1501 @@
+        else:
+            raise ValueError("Must provide one of `name` or `id`")
+        rec = {
+            "name":term.name,
+            "id": term.id,
+            "provider": self.name,
+            "composition": None,
+            "mass": self.get_mass_from_term(term)
+        }
+
+
+class GenericResolver(ModificationResolver):
+
+    def __init__(self, resolvers, **kwargs):
+        super(GenericResolver, self).__init__('generic', **kwargs)
+        self.resolvers = list(resolvers)
+
+    def load_database(self):
+        return None
+
+    def resolve(self, name=None, id=None, **kwargs):
+        defn = None
+        for resolver in self.resolvers:
+            try:
+                defn = resolver(name=name, id=id, **kwargs)
+            except (KeyError):
+                continue
+        if defn is None:
+            if name is None:
+                raise KeyError(id)
+            elif id is None:
+                raise KeyError(name)
+            else:
+                raise ValueError("Must provide one of `name` or `id`")
+        return defn
+
+
+class ModificationBase(TagBase):
+    '''A base class for all modification tags with marked prefixes.
+    '''
+
+    _tag_type = None
+    __slots__ = ('_definition', )
+
+    def __init__(self, value, extra=None, group_id=None):
+        super(ModificationBase, self).__init__(
+            self._tag_type, value, extra, group_id)
+        self._definition = None
+
+    @property
+    def definition(self):
+        if self._definition is None:
+            self._definition = self.resolve()
+        return self._definition
+
+    @property
+    def mass(self):
+        return self.definition['mass']
+
+    @property
+    def composition(self):
+        return self.definition.get('composition')
+
+    @property
+    def id(self):
+        return self.definition.get('id')
+
+    @property
+    def name(self):
+        return self.definition.get('name')
+
+    @property
+    def provider(self):
+        return self.definition.get('provider')
+
+    def _populate_from_definition(self, definition):
+        self._definition = definition
+
+    def _format_main(self):
+        return "{self.prefix_name}:{self.value}".format(self=self)
+
+    def _parse_identifier(self):
+        tokens = self.value.split(":", 1)
+        if len(tokens) > 1:
+            value = tokens[1]
+        else:
+            value = self.value
+        if value.isdigit():
+            id = int(value)
+            name = None
+        else:
+            name = value
+            id = None
+        return name, id
+
+    def resolve(self):
+        '''Find the term and return it's properties
+        '''
+        keys = self._parse_identifier()
+        return self.resolver(*keys)
+
+
+class FormulaModification(ModificationBase):
+    prefix_name = "Formula"
+
+    _tag_type = TagTypeEnum.formula
+
+    def resolve(self):
+        # The handling of fixed isotopes is wrong here as Pyteomics uses a different
```

Comment:
Not only is it wrong, but the formula is currently unsalvageable because the parser swallows all the square brackets.
`'Formula:[13C2][12C-2]H2N'` becomes `'13C212C-2H2N'`, where two numbers `2` and `12` are concatenated and the information is lost. Would there be a way to prevent the parser from consuming the `[` and `]` when they are actually parts of the data (formula) and not markup? Then it shouldn't be too hard to re-format the formula to be correctly parsed by pyteomics into a composition.

Reply:
Good catch and easily done. Thank you.